scripts/notebook-normalizer/qiskit_docs_notebook_normalizer/__init__.py

```diff
@@ -0,0 +1,221 @@
+    Path("learning").rglob("*.ipynb"),
+)
+
+# Qiskit's QuantumCircuit.draw() results in Jupyter outputting both a `text/html` and
+# `text/plain` entry. The HTML entry has pre-applied formatting that makes sense in
+# a Jupyter notebook, but renders horribly in our app:
+# https://github.com/Qiskit/qiskit/blob/df379876ba10d6f490a96723b6dbbf723ec45d7a/qiskit/visualization/circuit/text.py#L761-L769
+#
+# So, we instead should render the `text/plain` entry rather than `text/html`.
+CIRCUIT_DRAW_HTML_PREFIX = '<pre style="word-wrap: normal;white-space: pre;background: #fff0;line-height: 1.1;font-family: &quot;Courier New&quot;,Courier,monospace">'
+
+# Types for extracted images
+
+
+@dataclass
+class SvgImage:
+    data: str
+    filepath: Path
+
+    def write(self):
+        self.filepath.write_text(self.data)
+
+
+@dataclass
+class RasterImage:
+    data: bytes
+    filepath: Path
+
+    def write(self):
+        self.filepath.write_bytes(self.data)
+
+
+Image = SvgImage | RasterImage
+
+
+# Result types for normalization process
+@dataclass
+class NormalizationNeeded:
+    changes = True
+    nb: nbformat.NotebookNode
+    images: list[Image]
+
+
+@dataclass
+class AlreadyNormalized:
+    changes = False
+
+
+NormalizationResult = NormalizationNeeded | AlreadyNormalized
+
+
+def changes_made(result: NormalizationResult) -> TypeGuard[NormalizationNeeded]:
+    return result.changes
+
+
+def main():
+    """
+    Search for notebooks and extract image outputs if necessary.
+    """
+    parser = argparse.ArgumentParser(prog="Qiskit/documentation notebook normalization")
+    parser.add_argument("--check", action="store_true")
+    args = parser.parse_args()
+
+    problem_notebooks = []
+    for nb_path in NOTEBOOK_PATHS:
+        if is_hidden(nb_path):
+            continue
+
+        print(f"{nb_path}")
+        nb = nbformat.read(nb_path, 4)
+        images_folder = determine_image_folder(nb_path)
+
+        result = normalize_notebook(nb, images_folder, args.check)
+        if not changes_made(result):
+            continue
+
+        problem_notebooks.append(nb_path)
+        if args.check:
+            continue
+
+        ensure_exists_and_empty(images_folder)
+        for image in result.images:
+            image.write()
+        nbformat.write(result.nb, nb_path)
+        print(f"  Written notebook and {len(result.images)} image(s)")
+
+    if args.check and problem_notebooks:
+        print(
+            "\nThe following notebooks need normalizing:\n ",
+            "\n  ".join(map(str, problem_notebooks)),
+            "\nRun ./fix to fix them automatically.",
+        )
+        raise SystemExit(1)
+
+
+def normalize_notebook(
+    nb: nbformat.NotebookNode, image_folder: Path, check_only: bool = False
+) -> NormalizationResult:
+    """
+    Extracts images (converting if necessary) and returns an updated notebook.
+    """
+    images = []
+    change_made = False
+    for cell_index, cell in enumerate(nb.cells):
```

Comment:
This function has gotten pretty big. Maybe extract a helper function `normalize_cell`

Reply:
Good point, I tried splitting the specific normalization "passes" into their own functions – do you think that's enough? Or do you think it's better to have a per-cell function too?